### PR TITLE
KAFKA-7420: Global store surrounded by read only implementation

### PR DESCRIPTION
**Pull request by nizhikov**
Global store surrounded by read-only `KeyValueStore` implementation.
All methods that try to write into Store will throw `UnsupportedOperationException`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by nizhikov** [comment]
Hello, @mjsax! Can you take a look at my changes?

**Comment by nizhikov** [comment]
Failure unrelated.
Retest this, please

**Comment by nizhikov** [comment]
Failure unrelated.
Retest this, please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: Could we return a `ReadOnlyKeyValueStore` -- frankly speaking I am not sure, because it would not be backward compatible... \cc @guozhangwang WDYT?

**Review by mjsax — COMMENTED** [review]
Thanks for the PR @nizhikov! We should add a new `ProcessorContextImplTest` -- would be even better, if we test all existing methods, too, to close this testing gap.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (nizhikov) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: I think we can't return `ReadOnlyKeyValueStore` here.

From every usage of `getStateStore` I saw in the source code, the result is cast to `KeyValueStore`. 
And `ReadOnlyKeyValueStore` don't implement `KeyValueStore`.

`WordCountProcessorDemo.java` line 80: 
```
this.kvStore = (KeyValueStore<String, Integer>) context.getStateStore("Counts");
```

I think many users used the same technique to program their Processors. 
So users will obtain the runtime exception in the code that was correct in the previous version.

**Comment by nizhikov** [comment]
Failure unrelated.
Retest this, please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: Yes. That is what I meant by

> frankly speaking I am not sure, because it would not be backward compatible... 

It would be cleaner if we could return `ReadOnlyKeyValueStore`, but as a matter of fact, I also believe the most people cast to `KeyValueStore` -- not sure if we could argue that they wrote incorrect code. It was never documented...

**Comment by mjsax** [comment]
@nizhikov We stabilized a couple of test recently. Can you rebase this PR? This should help to get green builds.

**Comment by nizhikov** [comment]
@mjsax 

> We should add a new ProcessorContextImplTest

Please, tell me, what should be tested by this?

**Comment by nizhikov** [comment]
Failure unrelated.
Retest this, please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (kamalcph) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: `flush` is also a write operation. Can you throw `UnsupportedOperationException` here ?

Comment (nizhikov) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: Fixed.

**Comment by nizhikov** [comment]
@mjsax

> We should add a new ProcessorContextImplTest

Can you help me with this? What should be tested?

**Comment by mjsax** [comment]
We should test, that if a globalstore is returned from the context, calling any of the protected methods throws an exception.

**Comment by nizhikov** [comment]
@mjsax Test added. Please, review.

**Comment by nizhikov** [comment]
Tests passed

**Comment by nizhikov** [comment]
@mjsax Can you take a look at my changes?

**Comment by mjsax** [comment]
Call for second review. Anyof @guozhangwang @bbejeck @vvcephei

**Review by bbejeck — COMMENTED** [review]
Thanks for the contribution @nizhikov!  

Overall looks good, but I think we should get test coverage on the non-exception throwing methods from the `KeyValueStore` wrapper of the global store.

**Comment by nizhikov** [comment]
@bbejeck Thank you for the review.

I've added tests for non-exception methods of `KeyValueStore`.
Please, take a look.

**Comment by mjsax** [comment]
Build failed with a checkstyle error

**Review by bbejeck — APPROVED** [review]
@nizhikov thanks for the updated tests, LGTM pending the checkstyle fix

**Comment by nizhikov** [comment]
@mjsax 
@bbejeck 

Thanks for the review.

> Build failed with a checkstyle error
> LGTM pending the checkstyle fix

:man_facepalming: Sorry about that. Fixed

**Comment by nizhikov** [comment]
Tests passed. Can we merge this PR?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: nit: we usually format code differently and put `@Override` is its own line.

**Review by mjsax — COMMENTED** [review]
I was just about to merge this PR, but I suddenly have doubts about the correctness. For global stores, we do not dictate that it must be a `KeyValueStore` (only for `GlobalKTables`) if I am not wrong.

Thus, if somebody uses a custom store, the cast into a `KeyValueStore` in L79 would fail and we would crash.

For library provided stores, it's only possible to use a `KeyValueStore` though (`WindowStore` and `SessionStores` are not supported for the provided default implementations). However, users could also use custom implementation of `WindowStore` or `SessionStore` -- the problem is not with the interfaces, but with the implementation of the or built-in window/session store.

Thus, I think, we need to check the type of the global store, and wrap it accordingly if it implements one of the three store interfaces. The store type might also so unknown, and for this case we need to return the store as-is.

Does this make sense?

It would be great to add test for all those cases.

**Comment by nizhikov** [comment]
> Thus, I think, we need to check the type of the global store, and wrap it accordingly if it implements one of the three store interfaces. 

We should wrap in read only decorator instances that implements 3 following interfaces:

1. KeyValueStore
2. WindowStore
3. SessionStore

@mjsax , am I understand you correctly?

I will provide this fix and tests shortly.

**Comment by mjsax** [comment]
Yes. Check the type, for KeyValueStore do what the current PR does, and for Window/Session-Store apply the same pattern.

**Comment by nizhikov** [comment]
@mjsax I've implement your proposal. Please, review.

**Comment by nizhikov** [comment]
Tests are green.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: nit: please add `{ }` -- we always use them, even for single line blocks.

(@vvcephei You are our checkstyle expert -- can we add this to the checkstyle and let the build fail for missing `{ }` ?)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: nit: `ERR_MSG` -> `ERROR_MESSAGE` (avoid abbreviations to increase code readability)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: nit: this can be package private (what is more restrictive)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java]: nit: add `{ }`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java]: Why do we add `KeyValueStore`, but not `WindowStore`/`SessionsStore` ? If this can be any dummy, I would rename to `dummyStore` -- otherwise it might be confusing.

**Review by mjsax — COMMENTED** [review]
Thanks fo the update! Couple of nits.

Call for review @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: Yep! http://checkstyle.sourceforge.net/config_blocks.html#NeedBraces

Comment (nizhikov) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: fixed

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (nizhikov) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: fixed.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (nizhikov) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: fixed

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java
Comment (nizhikov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java]: fixed

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java
Comment (nizhikov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java]: This set can be replaced with empty.
Fixed

**Comment by nizhikov** [comment]
> Thanks for the update! Couple of nits.

Nits fixed.

**Comment by nizhikov** [comment]
Tests are green.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java]: Follow up PR: https://github.com/apache/kafka/pull/5954

**Comment by nizhikov** [comment]
Hello, @guozhangwang

Please, review and my merge my PR.

I have 2 approval from mjsax and bbejeck

**Comment by nizhikov** [comment]
@mjsax Do I need to improve this PR? Can you approve it?

**Review by mjsax — APPROVED** [review]

**Comment by nizhikov** [comment]
Hello @mjsax 
I got 2 OKs from you and bbejeck
Do I need one more review from guozhangwang?

No rush, I just want to know, how it's work in Kafka community

Can you merge this PR by yourself? :)

**Comment by mjsax** [comment]
@nizhikov Committers can merge PRs.

I would like to get a confirmation from @bbejeck because we did some mayor changes after he approved. (Or a second review from @guozhangwang @vvcephei).

We want two +1 before merging.

**Review by bbejeck — APPROVED** [review]
I took another pass over the updates LGTM

**Comment by nizhikov** [comment]
@bbejeck Thank you.

@mjsax  There are two +1. Let's merge it?! :)

**Comment by mjsax** [comment]
SGTM. There was a merge into `trunk` enabling `spotBugs` -- can you rebase this PR one more time to make sure the build is green (just as a guard to not break `trunk` if we merge). Thanks.

**Comment by nizhikov** [comment]
> can you rebase this PR one more time to make sure the build is green (just as a guard to not break `trunk` if we merge). Thanks.

Done

**Comment by nizhikov** [comment]
@mjsax 

> just as a guard to not break trunk if we merge

Tests are green.

**Comment by nizhikov** [comment]
Hello, @mjsax 
Is there anything that should be improved prior merge?

**Comment by nizhikov** [comment]
Hello, @mjsax 

Do you have a change do look at the tests results?
Can we merge PR?

Should I ask Bill Bejeck or Guozhang Wang to review and merge this patch?

**Comment by mjsax** [comment]
Thanks for the PR @nizhikov! Merged to `trunk`.

Please be a little patient. We try to get back to PRs as soon as we can, but it might take couple of days. :)

**Comment by nizhikov** [comment]
Thank you, @mjsax 

Sorry for pinging you too many times.
WIll try to avoid it in the next contributions.

**Pull request merged**
